docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,225 @@
+1. We will know we’ve succeeded when we can benchmark speed increases and successful image changes
+
+### Non-Goals/Future Work
+
+1. Extend the functionality to Windows nodes
+1. Optimization for efficiency and scalability
+1. A more complicated method of selecting a good candidate node such as incorporating manual forced prototype creation, which would perform better but take research to find the optimal method. 
+1. Optimization of the default time interval into a specific, best general default which would have to be researched
+
+## Proposal
+
+### User Stories
+
+#### As an operator of CAPZ, I want faster horizontal scaling on new node deployment
+
+People setting up CAPZ clusters will be able to have faster horizontal scaling because their Nodes’s OS images will be continually cached, and thus they will be able to avoid longer pull times and avoid having to create warm nodes and overprovision.
+
+#### As an operator, I want to be able to be more inline with security compliance as I spin up nodes
+
+As an operator I would like to be able to have my node’s OS image cache for the ability to avoid security alerts, cache security updates and patches so that I’ll be more security compliant and up to date when I have to spin up a new node. Otherwise, flags will be raised that a node is out of date on boot up before it finishes patching and a temporary security risk will exist.
+
+#### As an operator, I want to be able to have pods run quicker when scheduled
+
+As a cluster operator I want to bring up new nodes that already have the container images of the applications it will run pre-cached so that pods start running quicker when scheduled.
+
+#### Credits
+
+A big thank you to Jack Francis, Michael Sinz, and Amr Hanafi for their work on the Kamino project that used a similar Prototype Pattern with AKSEngine which inspired this feature request: jackfrancis/kamino: Kubernetes node clone factory (github.com)
+
+### Requirements (Optional)
+
+#### Functional Requirements
+
+None currently known to be needed
+
+##### FR1
+
+##### FR2
+
+#### Non-Functional Requirements
+
+##### NFR1
+The solution should increase the performance of horizontal scaling over time as people use this feature
+
+##### NFR2
+The solution should cause no increased risks with security or reliability
+
+### Implementation Details/Notes/Constraints
+
+The plan is to create a Controller which implements the feature with the Node Prototype Pattern as desired.
+
+These controller additions can be added to azuremachine_controller.go because that is where AzureMachines are being controlled and azuremachinepool_controller.go since that is where the AzureMachinePools are being controlled.
+
+An operator will be able to decide to turn the feature on or off with a flag on creation of a cluster, and then can alter the YAML file for AzureMachineTemplate and AzureMachinePool to customize how long they want the caching interval to be.
```

Comment:
Ended up switching the proposal from a flag to an environment variable since that seems like it will be easier to integrate without CAPI changes, added an example. Configuration of the caching interval is noted to be in the YAMLs for the AzureMachineTemplates and AzureMachinePools.